kaggle_environments/envs/lux_ai_2022/luxai2022/env.py

```diff
@@ -0,0 +1,618 @@
+                self.state.board.lichen_strains[indexable_positions] = factory.num_id
+            
+            self.state.board.lichen -= 1
+            self.state.board.lichen = self.state.board.lichen.clip(0)
+            self.state.board.lichen_strains[self.state.board.lichen == 0] = -1
+
+            # resources refining
+            for agent in self.agents:
+                factories_to_destroy: Set[Factory] = set()
+                for factory in self.state.factories[agent].values():
+                    factory.refine_step(self.env_cfg)
+                    factory.cargo.water -= self.env_cfg.FACTORY_WATER_CONSUMPTION
+                    if factory.cargo.water < 0:
+                        factories_to_destroy.add(factory)
+                for factory in factories_to_destroy:
+                    # destroy factories that ran out of water
+                    self.destroy_factory(factory)
+            # power gain
+            if is_day(self.env_cfg, self.state.real_env_steps):
+                for agent in self.agents:
+                    for u in self.state.units[agent].values():
+                        u.power = u.power + math.ceil(self.env_cfg.ROBOTS[u.unit_type.name].CHARGE * weather_cfg["power_gain_factor"])
+            for agent in self.agents:
+                for f in self.state.factories[agent].values():
+                    f.power = f.power + math.ceil(self.env_cfg.FACTORY_CHARGE * weather_cfg["power_gain_factor"])
+
+
+        # always set rubble under factories to 0.
+        self.state.board.rubble[self.state.board.factory_occupancy_map != -1] = 0
+
+        # rewards for all agents are placed in the rewards dictionary to be returned
+        rewards = {}
+        for agent in self.agents:
+            strain_ids = self.state.teams[agent].factory_strains
+            if failed_agents[agent]:
+                rewards[agent] = -1000
+            else:
+                agent_lichen_mask = np.isin(self.state.board.lichen_strains, strain_ids)
+                rewards[agent] = self.state.board.lichen[agent_lichen_mask].sum()
+
+        self.env_steps += 1
+        self.state.env_steps += 1
+        env_done = self.state.real_env_steps >= self.state.env_cfg.max_episode_length
+        dones = {agent: env_done or failed_agents[agent] for agent in self.agents}
+
+        # generate observations
+        obs = self.state.get_obs()
+        observations = {}
+        for k in self.agents:
+            observations[k] = obs
+
+        # log stats and other things
+        infos = {agent: {} for agent in self.agents}
+
+        if env_done:
+            self.agents = []
+
+        return observations, rewards, dones, infos
+
+    ### Game Logic ###
+    def add_unit(self, team: Team, unit_type, pos: np.ndarray):
+        unit = Unit(team=team, unit_type=unit_type, unit_id=f"unit_{self.state.global_id}", env_cfg=self.env_cfg)
+        unit.pos.pos = pos.copy()
+        # TODO allow user to specify how much power to give unit?
```

Comment:
I don't think this is necessary, they spawn in the factory so simple to pick some up on the way out